File: iOS/Sources/AppReveal/Network/NetworkObserver.swift
// Network traffic observation

import Foundation

#if DEBUG

/// Conform your network client to expose traffic to AppReveal.
public protocol NetworkObservable: AnyObject {
    var recentRequests: [CapturedRequest] { get }
    func addObserver(_ observer: NetworkTrafficObserver)
}

/// Receives network traffic events.
public protocol NetworkTrafficObserver: AnyObject {
    func didCapture(_ request: CapturedRequest)
}

@MainActor
final class NetworkObserverService: NetworkTrafficObserver {

    static let shared = NetworkObserverService()

    private let maxBufferSize = 200
    private(set) var capturedRequests: [CapturedRequest] = []
    private weak var observable: NetworkObservable?

    private init() {}

    func register(_ observable: NetworkObservable) {
        self.observable = observable
        observable.addObserver(self)
    }

    nonisolated func didCapture(_ request: CapturedRequest) {
        Task { @MainActor in
            capturedRequests.append(request)
            if capturedRequests.count > maxBufferSize {
                capturedRequests.removeFirst(capturedRequests.count - maxBufferSize)
            }
        }
    }

    func recentCalls(limit: Int = 50) -> [CapturedRequest] {
        Array(capturedRequests.suffix(limit))
    }

    func callDetail(id: String) -> CapturedRequest? {
        capturedRequests.first { $0.id == id }
    }

    func clear() {
        capturedRequests.removeAll()
    }
}

#endif

File: Android/appreveal/src/main/kotlin/com/appreveal/network/NetworkObserver.kt
package com.appreveal.network

/**
 * Implement on your network client to expose traffic to AppReveal.
 */
interface NetworkObservable {
    val recentRequests: List<CapturedRequest>

    fun addObserver(observer: NetworkTrafficObserver)
}

/**
 * Receives network traffic events.
 */
interface NetworkTrafficObserver {
    fun didCapture(request: CapturedRequest)
}

/**
 * Internal service that collects network traffic in a ring buffer.
 */
internal object NetworkObserverService : NetworkTrafficObserver {
    private const val MAX_BUFFER_SIZE = 200
    private val capturedRequests = LinkedHashMap<String, CapturedRequest>()
    private var observable: NetworkObservable? = null

    @Synchronized
    fun register(observable: NetworkObservable) {
        this.observable = observable
        observable.addObserver(this)
        observable.recentRequests.forEach(::didCapture)
    }

    @Synchronized
    override fun didCapture(request: CapturedRequest) {
        addOrUpdate(request.withRedactedHeaders())
    }

    @Synchronized
    fun addCall(request: CapturedRequest) {
        addOrUpdate(request.withRedactedHeaders())
    }

    @Synchronized
    fun updateCall(
        id: String,
        transform: (CapturedRequest) -> CapturedRequest,
    ) {
        val existing = capturedRequests[id] ?: return
        addOrUpdate(transform(existing).withRedactedHeaders())
    }

    @Synchronized
    fun recentCalls(limit: Int = 50): List<CapturedRequest> {
        val size = capturedRequests.size
        val fromIndex = maxOf(0, size - limit)
        return capturedRequests.values.toList().subList(fromIndex, size).toList()
    }

    @Synchronized
    fun callDetail(id: String): CapturedRequest? = capturedRequests[id]

    @Synchronized
    fun clear() {
        capturedRequests.clear()
    }

    private fun addOrUpdate(request: CapturedRequest) {
        capturedRequests[request.id] = request
        while (capturedRequests.size > MAX_BUFFER_SIZE) {
            val oldest = capturedRequests.keys.firstOrNull() ?: return
            capturedRequests.remove(oldest)
        }
    }
}
